pkg/condition/condition.go

```diff
@@ -0,0 +1,171 @@
+
+	patcher "github.com/openshift/windows-machine-config-operator/pkg/patch"
+	"github.com/openshift/windows-machine-config-operator/pkg/retry"
+)
+
+const (
+	upgradeableTrueReason       = "upgradeIsSafe"
+	upgradeableTrueMessage      = "The operator is safe for upgrade"
+	upgradeableFalseReason      = "upgradeIsNotSafe"
+	upgradeableFalseMessageBase = "The operator is currently processing sub-components"
+	// OperatorConditionName is an environment variable set by OLM identifying the operator's OperatorCondition CR
+	OperatorConditionName = "OPERATOR_CONDITION_NAME"
+)
+
+var (
+	// ocName is the name of the OLM-managed OperatorCondition resource. Empty if operator is not OLM-managed
+	ocName string
+	// busyControllers is a representation of a set that holds the names of controllers that are currently reconciling
+	busyControllers map[string]struct{}
+	// resourceLock is a mutex lock to ensure thread-safe access of shared entities
+	resourceLock sync.Mutex
+)
+
+// init runs once, initializing global variables
+func init() {
+	ocName = getResourceName()
+	busyControllers = make(map[string]struct{})
+}
+
+// exported methods
+
+// MarkAsFree removes the given controller from the set of busy controllers, sets the Upgradeable condition to True
+// to allow OLM to perform operator upgrades
+func MarkAsFree(c client.Client, watchNamespace, controllerName string) error {
+	resourceLock.Lock()
+	defer resourceLock.Unlock()
+
+	// Remove given controller from busy controllers set
+	delete(busyControllers, controllerName)
+
+	// At least one controller is still busy, meaning Upgradeable condition should remain False. Do not update value.
+	if len(busyControllers) > 0 {
+		return nil
+	}
+	return set(c, watchNamespace, operators.Upgradeable, meta.ConditionTrue, upgradeableTrueReason,
+		upgradeableTrueMessage)
+}
+
+// MarkAsBusy marks the given conroller as busy and sets the Upgradeable condition to False
+// to prevent operator upgrades by OLM
+func MarkAsBusy(c client.Client, watchNamespace, controllerName string) error {
+	resourceLock.Lock()
+	defer resourceLock.Unlock()
+
+	// If given controller is already marked as busy, no need to update Upgradeable condition
+	if _, exists := busyControllers[controllerName]; exists {
+		return nil
+	}
+
+	// Add given controller to busy controllers set. struct{} used as value since it takes up 0 bytes
+	busyControllers[controllerName] = struct{}{}
+
+	// More than one controller is busy, meaning Upgradeable condition is already False. No need to update value.
+	if len(busyControllers) > 1 {
+		return nil
+	}
+	return set(c, watchNamespace, operators.Upgradeable, meta.ConditionFalse, upgradeableFalseReason,
+		fmt.Sprintf("%s: %s", upgradeableFalseMessageBase, controllerName))
+}
+
+// Validate checks that the given condition is present and proper within the given list
+func Validate(conditions []meta.Condition, conditionType string, expectedStatus meta.ConditionStatus) bool {
```

Comment:
Don't see this used outside of the package in this commit, maybe its used in a later commit

Reply:
yep was used in test commit

Reply:
Not sure why this and all its associated changes leaked into this commit. This belongs to `[wmco] Opt in to Upgradeable operator condition`. @sebsoto was your comment asking for this to be moved to this commit?